pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,254 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	fileutils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+)
+
+// PatchEnvelopes is a structure representing
+// Patch Envelopes exposed to App instances via metadata server
+// for more info check docs/PATCH-ENVELOPES.md
+// Must be created by calling NewPatchEnvelopes()
+//
+// Internally, PatchEnvelopes structure stores envelopes which
+// come from EdgeDevConfig parsed by zedagent. This envelopes contains
+// both inline binary artifacts which are ready to be downloaded by app instances
+// and volume references, which are handled by volumemgr. So PatchEnvelopes struct has
+// VolumeStatusCh and PatchEnvelopeInfoCh to process volumes and patch envelopes.
+// NewPatchEnvelopes() starts coroutine processMessages which reads from both channels
+// and stores latest state of patch envelopes exposed to app instances via Get()
+//
+// That means that VolumeStatusCh and PatchEnvelopeInfoCh should not be read from anywhere
+// except processMessages(). so there is no deadlocks.
+type PatchEnvelopes struct {
+	sync.Mutex
+	VolumeStatusCh      chan PatchEnvelopesVsCh
+	PatchEnvelopeInfoCh chan []types.PatchEnvelopeInfo
+
+	envelopes        []types.PatchEnvelopeInfo
+	completedVolumes []types.VolumeStatus
+	log              *base.LogObject
+}
+
+// PatchEnvelopesVsChAction specifies action for patch envelope volume status
+type PatchEnvelopesVsChAction uint8
+
+const (
+	// PatchEnvelopesVsChActionDelete -- delete reference with volume
+	PatchEnvelopesVsChActionDelete PatchEnvelopesVsChAction = iota
+	// PatchEnvelopesVsChActionPut -- create or update reference with volume
+	PatchEnvelopesVsChActionPut
+)
+
+// PatchEnvelopesVsCh is a wrapper for VolumeStatus specifying action
+// for external Patch Envelope
+type PatchEnvelopesVsCh struct {
+	Vs     types.VolumeStatus
+	Action PatchEnvelopesVsChAction
+}
+
+// NewPatchEnvelopes returns PatchEnvelopes structure and starts coroutine
+// to process messages from channels. Note that we create buffered channels
+// to avoid unbounded processing time in writing to channel
+func NewPatchEnvelopes(log *base.LogObject) *PatchEnvelopes {
+	pe := &PatchEnvelopes{
+		VolumeStatusCh:      make(chan PatchEnvelopesVsCh, 1),
+		PatchEnvelopeInfoCh: make(chan []types.PatchEnvelopeInfo, 1),
+
+		log: log,
+	}
+
+	go pe.processMessages()
+
+	return pe
+}
+
+// Get returns list of Patch Envelopes available for this app instance
+func (pes *PatchEnvelopes) Get(appUUID string) types.PatchEnvelopeInfoList {
+	var res []types.PatchEnvelopeInfo
+
+	for _, envelope := range pes.envelopes {
+		for _, allowedUUID := range envelope.AllowedApps {
+			if allowedUUID == appUUID {
+				res = append(res, envelope)
+				break
+			}
+		}
+	}
+
+	return types.PatchEnvelopeInfoList{
+		Envelopes: res,
+	}
+}
+
+func (pes *PatchEnvelopes) processMessages() {
```

Comment:
Please add a comment that processMessages and the functions it calls can not block hence they can not send on any channels.

Reply:
added comment :)